Preparatory work for adding realm default settings UI

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
- Refactor display settings frontend code to use classes instead of IDs.
- Refactor notification settings frontend code to use classes instead of IDs.
- Rename the notification sound elements.

<!-- How have you tested? -->


 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on static/templates/settings/display_settings.hbs:
```diff
@@ -1,39 +1,39 @@
                     {{> ../help_link_widget link="/help/change-default-view" }}
                 </label>
-                <select name="default_view" id="default_view">
+                <select name="default_view" class="default_view">
```
Do we want to extend these names to be something more like `setting_default_view` to namespace them a bit better, while we're changing this?

sahil839 replied:
Yes we can do this. Would be better in a separate commit as we would need to change other settings using `settings_checkbox`.
Do we want realm settings elements to be renamed also, like `realm_invite_required`?

timabbott commented:
This refactor looks reasonable to me; make sure you do manual testing carefully with the final version of this, as it's definitely the kind of change that tends to end up breaking some widgets.

sahil839 commented:
Yes, I have manually tested all the settings and have also checked the changes using `grep`.

sahil839 commented:
There are few elements which are used in JS code, but I cannot find them elsewhere. I suspect that those statements should be removed. I would look into that and update here so we do not merge anything buggy.

sahil839 commented:
@timabbott Added two commits to fix unnecessary code as mentioned above. 
Waiting for the reply whether to also rename the realm settings elements since they also use `settings_checkbox`

sahil839 commented:
@timabbott added the commits to rename the classes of all the setting elements except those using `settings_checkbox` because realm settings also use that module, wanted to confirm before renaming them. Have only updated classes and not names or ids because those are not used to access the elements.

timabbott commented:
I merged the first couple commits.  This commit has a problem:

> templates: Do not include '_' in prefix for settings_checkbox.

Some of our `settings_checkbox` references don't pass the `prefix` parameter, at last via this at the end of this branch:

```
$ git grep -A7 settings_checkbox
static/templates/settings/account_settings.hbs:                {{> settings_checkbox
static/templates/settings/account_settings.hbs-                  setting_name="presence_enabled"
static/templates/settings/account_settings.hbs-                  is_checked=user_settings.presence_enabled
static/templates/settings/account_settings.hbs-                  label=settings_label.presence_enabled
static/templates/settings/account_settings.hbs-                  help_link="/help/status-and-availability"}}
static/templates/settings/account_settings.hbs-            </div>
static/templates/settings/account_settings.hbs-        </div>
static/templates/settings/account_settings.hbs-
```

I think it'd be a reasonable change to make the `prefix` parameter required, but it'll require some refactoring.

sahil839 commented:
Ah, I missed this somehow. Will it be fine to have all the them prefixed with `id` like we do for organization settings.

sahil839 commented:
@timabbott  Actually, we can just add `user` as prefix for the presence_enabled also as this setting is also be included in the realm-level defaults (missed it previously because it is on the account settings page). And all other settings_checkboxes references have prefix after the changes in this PR. Do we also need to add any config rule or something like that to make sure future references of `settings_checkbox` have prefix parameter, though not sure how to do that.

sahil839 commented:
Not ready for next review yet. Need to do some manual testing.

sahil839 commented:
@timabbott ready for review. Updated the PR with following changes-
- Added a commit to change the `presence_enabled` elements have `user` prefix in ids.
- I removed the commit changing the prefix to be passed as "id" instead of "id_" as I added that only to ensure we can pass same prefix to the topmost container and the checkboxes. But after moving forward with the actual work for adding UI for realm-level settings in different branch, I realised we would need a new template file for rendering them with the new files having top-most div and then using `display_settings` as common template. So, I just moved those commits in this PR and now we do not need to change the prefix form. Though we have now covered all the `settings_checkbox` references to have prefix now (as I missed `presence_enabled` setting previously), I have removed that commit for now as future additions of `settings_checkbox` may miss the prefix and we would require to handle those by adding some restriction to not allow to do so.
- Also added separate commits for adding prefixes to checkboxes ids.
- I have not added the `setting_` prefix to checkboxes for now as they are used for organization settings also, so just wanted to discuss it first. And `notification_setting_checkbox` is used here only but it breaks the live-update code then.
- Also the wildcard mentions checkbox for streams which is always disabled now has `user_` id after adding the prefix, but I think we can ignore that.

timabbott commented:
Merged, after a small fix to `demote_inactive_streams` having been

@aryanshridhar @ganpa3 can both of you do a round of manual testing of these settings pages to make sure every element works as intended?  This is a pretty involved migration and independent testing to make sure we didn't break anything would be really valuable.

timabbott commented:
I think your notes above are the right decisions on all of those details.  I agree we should aim to rename the classes for those, checkboxes but doing so in a follow-up project that carefully manages the issue for organization settings makes sense.

sahil839 commented:
I missed to mention one thing - There is a bug in live update of notification settings. Only checkbox settings are updated and dropdown settings do not.

timabbott commented:
@sahil839 OK -- we should prioritize fixing that.  Is there an issue or PR tracking that?

sahil839 commented:
There is not any issue/PR but I will soon open a PR.
